[Copilot No. Series] Number series copilot

DmitryKatson opened the pull request:
<!-- Thank you for submitting a Pull Request. If you're new to contributing to BCApps please read our pull request guideline below
* https://github.com/microsoft/BCApps/Contributing.md
-->
#### Summary <!-- Provide a general summary of your changes -->

Introducing the first [draft] version of a "Number Series Copilot" for Business Central, an innovative feature designed to simplify the creation of number series through natural language commands. This allows users to effortlessly set up number series for specific entities like customers, vendors, or even across all entities, by simply describing their needs or specifying a format example.

Current version is intent to create number series for NEW company, where no Number Series exist yet. 

https://github.com/microsoft/BCApps/assets/29101017/95b0168b-70ef-40f9-9f8b-4ba2db34b46e

This flow was tested on the `Set up numbers series for the new company` like user prompts

Please follow this diagram (green flow), to get the idea of the ongoing process.

![diagram-export-16-03-2024-13_39_19](https://github.com/microsoft/BCApps/assets/29101017/9b8c5484-4942-4897-9d4e-572d93cf609a)


#### Key ideas:
- Azure OpenAI Tools calling is used to understand the user intent
- AL is used to gather data from BC and build tool response
- The new company flow requires about 120 different number series to be generated.  Nor GPT-4, not GPT 3.5 turbo (latest) where possible to generate accurate json array with 1 go
- Thus, instead of 1 LLM call to build 120 number series, the tool response is been chunked, to generate not more than 10 number series in one go
- All LLM generations results (12 in total, each requires 5-9 sec of GPT 3.5 time) then been combined in 1 answer, which is presented to the user

#### Setup:
- Use `No. Series with Copilot Setup` to set up Azure Open AI connection and prompts
- Prompts should be provided separately (using other channels) due to security reasons
- Model used: `gpt-35-turbo-1106`
<img width="1406" alt="image" src="https://github.com/microsoft/BCApps/assets/29101017/d4f88f87-e627-41a2-903f-1435f873fd93">


#### Work Item(s) <!-- Add the issue number here after the #. The issue needs to be open and approved. Submitting PRs with no linked issues or unapproved issues is highly discouraged. -->

Fixes #659
Fixes [AB#503148](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/503148)

pri-kise commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesExt.PageExt.al:
```diff
@@ -0,0 +1,60 @@
+pageextension 324 "No. Series Ext" extends "No. Series"
```
`Ext` could be a valid AppSource affix of a BC partner.
We should avoid three character affixes on Extension Objects to avoid collision with object names.

grobyns replied:
Ext has not been registered by itself but is part of some longer affixes. Now that we have namespaces though the risk of conflicts is negligeable.

JesperSchulz commented on src/Business Foundation/App/NoSeriesCopilot/src/Register/NoSeriesCopilotInstall.Codeunit.al:
```diff
@@ -0,0 +1,15 @@
+namespace Microsoft.Bank.Reconciliation;
```
That namespace doesn't look right.

DmitryKatson replied:
Removed for now

JesperSchulz commented on src/Business Foundation/App/NoSeriesCopilot/src/Register/NoSeriesCopilotInstall.Codeunit.al:
```diff
@@ -0,0 +1,15 @@
+namespace Microsoft.Bank.Reconciliation;
+
+codeunit 290 "No. Series Copilot Install"
```
You've got 2 of these codeunits 😉

DmitryKatson replied:
Yep, good that we have 2nd vision ;) Removed

JesperSchulz commented on src/System Application/App/Json/app.json:
```diff
@@ -0,0 +1,39 @@
+{
```
This code is already part of a separate PR.

DmitryKatson replied:
Yes, but how should i separate those?

JesperSchulz replied:
Thinking it through, I guess the easiest is that we get 716 and 719 processed. Once they're in, you can remove them from this DRAFT - and eventually take it out of DRAFT mode. There's a reason it's a DRAFT, right ;-)

JesperSchulz commented on src/System Application/App/AI/src/Azure OpenAI/AOAIOperationResponse.Codeunit.al:
```diff
@@ -20,6 +20,7 @@ codeunit 7770 "AOAI Operation Response"
         Success: Boolean;
```
These changes are part of https://github.com/microsoft/BCApps/pull/719

DmitryKatson replied:
Yes, but how should i separate those?

JesperSchulz replied:
We process it and then you can remove it.

darjoo commented:
Here are my notes/questions I have from our sync :)

- Action visibility
    - Addition to SaaS check, we need to limit CA and potentially any other languages. This is not necessary to add now but I'll add the code snippets below. CA is due to language requirement that both EN & FR are supported. Languages wise, we've typically stuck with EN first as Production Ready Previews and slowly roll out other languages as the language assessment team gives us the green light
- Use prompt guides
    - For the prompt dialog, using prompt guides can give the users an easy way to select premade prompts. The ones that you think are most likely to be used by them. I'm not sure if the docs are ready yet about them, but they will be mentioned in the Launch Event for What's New: Extending Copilot with AL
- There's a check for whether input against the max tokens, and it just exits if it's greater than the max tokens allowed. We should at least give an actionable error or see if it can be modified to always allow even if it's multiple requests
- When the generation comes back with a response, would it be beneficial for `AOAIChatMessages` to give you say, `GetLastTool` or `GetLastFunction` when it's a tool/function and provide you the `Name:Text; Arguments: JsonArray` (for example) instead of having to parse `GetLastMessage`?
- I'll add a method in `AOAIChatMessages` to clear the existing tools. So you don't have to loop through and delete all tools.
- In your `BuildGenerateNewNumberSeries` function, it runs tokencount for each line. I'll be improving that, so you can put in `AOAIChatMessages` directly and it'll count everything that's being sent.
- Filtering on name with `* Setup` for tables is risky as we can't guarantee they always end with `Setup`
- GPT3.5 has a context window of 16k while GPT4 has 32k and turbo 128k, would it be better to switch? 
    - This you had answered already, hallucination can be pretty bad with generating more tables. So there isn't a point in using GPT4
- Does generation of ~10 tables always work?
    - This you had also answered, it doesn't as such it does retries and completely fails if it doesn't generate all.
    - We should allow partial generation and then allow the the user to request another generation on the ungenerated ones, or a full regenerate.
- How long does it take to run? As running 10+ requests will take time, including retries that makes it worse.
    - Takes about a minute. This might be fine as this is an admin scenario where they're setting the system up. And this saves them a lot more time in the long run.
- Retry logic, what if 1 out of 10 fails?
    - Similar qs to the generation of ~10 tables always work. You mentioned that all of them fail. We should allow partial generation if it makes sense.

Thank you so much for going through the code and all! Looking forward to when this sees the light of day in the product!

JesperSchulz commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,917 @@
+        TablesPromptList.Add('Area: ' + RemoveTextPart(TableMetadata.Caption, ' Setup') + ', TableId: ' + Format(TableMetadata.ID) + ', FieldId: ' + Format(Field."No.") + ', FieldName: ' + RemoveTextPart(Field.FieldName, ' Nos.'));
+    end;
+
+    local procedure SetFilterOnNoSeriesFields(var TableMetadata: Record "Table Metadata"; var Field: Record "Field")
```
I'm not a fan of this way of finding fields which set up No. Series. While it's a best practice to call those fields something with Nos., I don't think we can/should rely on that. Maybe we can use the following filter in some way?
`Field.SetRange(RelationTableNo, Database::"No. Series");`
Now you would filter on all fields which have a relationship to No. Series. Given that you filter on Setup tables, that would be setup of No. Series, I assume.

I'm not super happy with the "SetFilterOnSetupTables" procedure though, as again we probably cannot rely on the fact that all setup tables end on Setup. But I haven't found a solution to that yet :-)

duichwer replied:
Especially there are affixes.
We mostly use prefix for our objects, but I know other partners that are using suffixes and therefore those setup tables won't be found.

Normally setup tables should have a primary key with Code[10].

Having an event to subscribe could be a solution but I'd prefer if we find some logic instead.

JesperSchulz commented:
#716 was merged 🥳 Please pull from main and discard this PRs changes to the JSON module.

DmitryKatson commented:
[This](https://github.com/microsoft/BCApps/pull/764/commits/6546377bd864ef08dd8aff197a2b86edc41f2ef8) commit adds support to the **change** number series flow. 

https://github.com/microsoft/BCApps/assets/29101017/3882800b-0fea-4e1b-92aa-605c274746c0

To setup this flow you need to paste required prompts into the `No. Series with Copilot Setup` `Tool 2` group fields.

<img width="696" alt="image" src="https://github.com/microsoft/BCApps/assets/29101017/3c3ee591-9e61-4585-8380-e7c7ed0f6947">

DmitryKatson commented:
> #716 was merged 🥳 Please pull from main and discard this PRs changes to the JSON module.

Great! I did the next steps:
1) My forked repo on the GitHub > Sync Fork
2) My local repo > Pull

Now I see this picture in my local repo

<img width="611" alt="image" src="https://github.com/microsoft/BCApps/assets/29101017/910c8033-322c-43ae-b891-5b773c7f9884">

JesperSchulz commented:
> > #716 was merged 🥳 Please pull from main and discard this PRs changes to the JSON module.
> 
> Great! I did the next steps:
> 
> 1. My forked repo on the GitHub > Sync Fork
> 2. My local repo > Pull
> 
> Now I see this picture in my local repo
> 
> <img alt="image" width="611" src="https://private-user-images.githubusercontent.com/29101017/316075684-910c8033-322c-43ae-b891-5b773c7f9884.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMjg3ODgsIm5iZiI6MTcxMTEyODQ4OCwicGF0aCI6Ii8yOTEwMTAxNy8zMTYwNzU2ODQtOTEwYzgwMzMtMzIyYy00M2FlLWI4OTEtNWI3NzNjN2Y5ODg0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIyVDE3MjgwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUyYzg3MjBlMTY1NjdjNWZmY2UxMTY5NWVlMzQyZjljOGZlYzVjYzA1N2Y5OWQxZmE2MDdhYmVjMDcyOWRjMjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9KSglYrAdyODgkmJM4ePaq46_8fjes65KNafc6mDw0I">

And the changes in your PR are gone! That's 1 down, 1 more to go. And then we get to the meaty parts of your work 😎

darjoo commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,1087 @@
+        MaxToolResponseTokenLength, ExpectedNoSeriesCount, i : Integer;
+    begin
+        // remove the tools from the chat messages, as they are not needed anymore
+        for i := 1 to AOAIChatMessages.GetTools().Count do
```
https://github.com/microsoft/BCApps/pull/824
The above PR will allow you to remove all the tools more easily.
`AOAIChatMessages.ClearTools();`

DmitryKatson commented:
@darjoo i found a bug in the prompt dialog rendering.
In the content area if the field has dynamic visibility, like here `IsResponseTextVisible`, when it has value and visible, the value is not shown in the screen. If the visibility is static, all fine.

```
        area(Content)
        {
            field(ResponseText; ResponseText)
            {
                Caption = 'AI Response';
                MultiLine = true;
                ApplicationArea = All;
                ShowCaption = false;
                Visible = IsResponseTextVisible;
                Editable = false;
                Enabled = false;
            }
            part(ProposalDetails; "No. Series Proposal Sub")
            {
                Caption = 'No. Series proposals';
                ShowFilter = false;
                ApplicationArea = All;
                Editable = false;
                Enabled = true;
                Visible = IsProposalDetailsVisible;
                SubPageLink = "Proposal No." = field("No.");
            }
        }
```

DmitryKatson commented:
_New milestone done_
Apply generated number series. 

This will save generated number series to the no series table AND apply to the Setup tables. 


https://github.com/microsoft/BCApps/assets/29101017/269fdb2b-c54b-4191-92aa-ce91eb0667a1

hoovercj commented:
> @darjoo i found a bug in the prompt dialog rendering. In the content area if the field has dynamic visibility, like here `IsResponseTextVisible`, when it has value and visible, the value is not shown in the screen. If the visibility is static, all fine.

@DmitryKatson  [per our documentation](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/properties/devenv-visible-property#dynamic-visibility-of-controls), Dynamic Visibility only works on groups, parts, and actions. It does not work on fields.

The workaround is to wrap the field in a group with `showCaption = false`, though this has other visual implications that may not be desired.

![image](https://github.com/microsoft/BCApps/assets/2957173/f7fb19cd-cb8a-4f7e-ae6c-7fc8457972ff)

DmitryKatson commented:
> > @darjoo i found a bug in the prompt dialog rendering. In the content area if the field has dynamic visibility, like here `IsResponseTextVisible`, when it has value and visible, the value is not shown in the screen. If the visibility is static, all fine.
> 
> @DmitryKatson [per our documentation](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/properties/devenv-visible-property#dynamic-visibility-of-controls), Dynamic Visibility only works on groups, parts, and actions. It does not work on fields.
> 
> The workaround is to wrap the field in a group with `showCaption = false`, though this has other visual implications that may not be desired.
> 
> ![image](https://private-user-images.githubusercontent.com/2957173/318713890-f7fb19cd-cb8a-4f7e-ae6c-7fc8457972ff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwNTMyMTgsIm5iZiI6MTcxMjA1MjkxOCwicGF0aCI6Ii8yOTU3MTczLzMxODcxMzg5MC1mN2ZiMTljZC1jYjhhLTRmN2UtYWU2Yy03ZmM4NDU3OTcyZmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDJUMTAxNTE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWVjMWEzYjJlOTA0ZjM0MDQ2NDhiOWU2MzkzNTRmY2U1MGQ5MTQzM2Y4NTc5MjcxYzAwNjY3ZDYzMjIzZGViMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.GIKN5BPeGtNNsMdoyUbtXWcw6nBgFg72GuvgdzHBCtQ)

Thanks! That worked

DmitryKatson commented:
@JesperSchulz @darjoo 
Just finished this PR. I think it's ready for review. 
- Added new UI components: text placeholder and prompt guides
- Made possible to edit copilot proposals, before saving
- Refactored code: spitted big `No. Series Copilot Impl.` into tool calling modules
- Refined prompts, for better results and less retries. (Still need some way to pass to you)

In general, it works as I expected: user can create new and modify existing no series with AI :)

https://github.com/microsoft/BCApps/assets/29101017/1c080196-02fe-40b3-bde4-f7404571f760

JesperSchulz commented:
> @JesperSchulz @darjoo Just finished this PR. I think it's ready for review.
> 
> * Added new UI components: text placeholder and prompt guides
> * Made possible to edit copilot proposals, before saving
> * Refactored code: spitted big `No. Series Copilot Impl.` into tool calling modules
> * Refined prompts, for better results and less retries. (Still need some way to pass to you)
> 
> In general, it works as I expected: user can create new and modify existing no series with AI :)
> 
>  new-modify-no-series-with-prompt-guides.mov

Wow, we better get busy then! But as discussed, having created the PR is just the very beginning! Now all the fun begins to live up to our responsible AI rules 🙂 We'll be in touch! In very close touch 😉

DmitryKatson commented:
Not sure if i did everything correctly. But due to the latest Tools Calling Implementation refactoring, this PR needs to be reimplemented.

DmitryKatson commented:
@darjoo @JesperSchulz Finished the refactoring, to make it compatible with latest AOAI module changes. Ready for review.

JesperSchulz commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesExt.PageExt.al:
```diff
@@ -0,0 +1,63 @@
+pageextension 324 "No. Series Ext" extends "No. Series"
```
On the note of namespaces: we need those ;-)

JesperSchulz commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,509 @@
+codeunit 324 "No. Series Copilot Impl."
+{
+    var
+        IncorrectCompletionErr: Label 'Incorrect completion. The property %1 is empty';
```
Placeholders must have a comment.

JesperSchulz commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,509 @@
+        SystemPromptTxt: SecretText;
+        CompletePromptTokenCount: Integer;
+        Completion: Text;
+        TokenCountImpl: Codeunit "AOAI Token";
```
You should order all of your variables properly, or CodeCop will be on your neck ;-)

JesperSchulz commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,509 @@
+    local procedure GenerateNoSeries(SystemPromptTxt: SecretText; InputText: Text): Text
+    var
+        AzureOpenAI: Codeunit "Azure OpenAi";
+        AOAIDeployments: Codeunit "AOAI Deployments";
```
Unused. You should enable the CodeCops to get all of minor issues sorted out. Won't mention further of those.

DmitryKatson commented:
1 min 20 sec. 
This time you need now to configure No Series with Copilot for the whole new company.

https://github.com/microsoft/BCApps/assets/29101017/b27722ae-1b8a-42bb-a354-57fd42fcf116

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,509 @@
+
+    procedure FeatureName(): Text
+    begin
+        exit('Number Series with AI');
```
Label? If it should not be translated lock

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesExt.PageExt.al:
```diff
@@ -0,0 +1,63 @@
+
+    var
+        CopilotActionsVisible: Boolean;
+        NumberSeriesWithAILbl: label 'NumberSeriesWithAI', Locked = true;
```
Casing 'label'

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,509 @@
+        exit(RandIntInRange(33, 126)); // ASCII: ! (33) to ~ (126)
+    end;
+
+    local procedure RandIntInRange("Min": Integer; "Max": Integer): Integer
```
Max and Min parameters with " and than used without ". Should be consistend, also matches keywork, could be changed to MinInt, MaxInt.

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,509 @@
+
+    local procedure CheckIntegerPropertyExistAndCheckIfNotEmpty(propertyName: Text; var Json: Codeunit Json)
+    var
+        value: Integer;
```
Please rename to PropertyValue

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesProposal.Page.al:
```diff
@@ -0,0 +1,189 @@
+                ShowFilter = false;
+                ApplicationArea = All;
+                Editable = true;
+                Enabled = true;
```
Enabled default value true, no need to define it separatly

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesProposal.Table.al:
```diff
@@ -0,0 +1,46 @@
+table 391 "No. Series Proposal"
+{
+    TableType = Temporary;
+
```
DataClasification missing

JesperSchulz replied:
This is a temporary table. No classification needed, as data isn't stored.

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesProposal.Table.al:
```diff
@@ -0,0 +1,46 @@
+        Result: Text;
+    begin
+        Rec.CalcFields("Input Text");
+        Rec."Input Text".CreateInStream(InStr, TextEncoding::UTF8);
```
This code will throw an error if Stream is not populated, which is possible since there is no check on call GetInputText neither in this procedure

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesProposalLine.Table.al:
```diff
@@ -0,0 +1,102 @@
+table 392 "No. Series Proposal Line"
+{
+    TableType = Temporary;
```
DataClasification on all tables, wont' comment further this one

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/Tools/NoSeriesCopToolsImpl.Codeunit.al:
```diff
@@ -0,0 +1,243 @@
+    Access = Internal;
+
+    var
+        CustomPatternsPlaceholderLbl: label '{custom_patterns}', Locked = true;
```
'label' check on all places casing

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/Tools/NoSeriesCopToolsImpl.Codeunit.al:
```diff
@@ -0,0 +1,243 @@
+        exit(Text);
+    end;
+
+    procedure RemoveTextPart(Text: Text; Part: Text): Text
```
![image](https://github.com/microsoft/BCApps/assets/59221826/9e07c147-aebc-41a8-876c-19202ea33fd6)

So far only place where this case is existing is on Lookup triggers, everywhere else should be prefixed or affixed with something meaningful, Text: Text) : Text is just too confusing...

StefanSosic commented on src/Business Foundation/App/NoSeriesCopilot/src/Upgrade/NoSeriesCopilotUpgrTags.Codeunit.al:
```diff
@@ -0,0 +1,11 @@
+
+    procedure GetImplementationUpgradeTag(): Code[250]
+    begin
+        exit('MS-659-AddImplementationExtensibility-20240304 '); //659 is the id of the issue https://github.com/microsoft/BCApps/issues/659
```
trailing space

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/Permissions/NoSeriesCopilotObjects.permissionset.al:
```diff
@@ -0,0 +1,23 @@
+        codeunit "No. Series Copilot Register"=X,
+        codeunit "No. Series Copilot Upgr. Tags"=X,
+        page "No. Series Copilot Setup"=X,
+        tabledata "No. Series Copilot Setup"=RIMD;
```
the *objects permission sets should not contain tabledata. These permissionsets exist as building blocks for those customers that do not use <type> * = X in their permissionsets which means they need a way to run the internal objects, and this is it. 
The tabledata permissions should be split by usage scenario, but the defaults we use are Read, View, Edit or Admin 
 where not all are required and we typically have Read + 1 or two more. Each subsequent level includes the previous. As some of these might seem ambiguous, here's how I see them:
 
 Read = only r or R.
 View = ability to directly view the data on screen but no direct edit ability (optional indirect imd).
 Edit = direct edit on pages

JesperSchulz replied:
Sorry, this one is my bad. I had the permission set generated to please CodeCop, and then I accidentally checked in the permission set before actually getting this right. I will revisit the permissions!

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+using System.Text.Json;
+
+codeunit 324 "No. Series Copilot Impl."
+{
```
"Impl." codeunits should always have Access = Internal. 
If some of this should be public you should add a facade codeunit.

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+    var
+        FeatureTelemetry: Codeunit "Feature Telemetry";
+        NoSeriesCopilotRegister: Codeunit "No. Series Copilot Register";
+        NoSeriesCopilotImpl: Codeunit "No. Series Copilot Impl.";
```
why this variable? as far as i can tell it is only used to get the FeatureName so you could just call this directly

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+        NoSeriesCopilotRegister: Codeunit "No. Series Copilot Register";
+        NoSeriesCopilotImpl: Codeunit "No. Series Copilot Impl.";
+        AzureOpenAI: Codeunit "Azure OpenAI";
+        NoSeriesCopilot: Page "No. Series Proposal";
```
Why call the page No. Series Proposal but the variable NoSeriesCopilot?

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+        NoSeriesCopilot.Run();
+    end;
+
+    procedure Generate(var NoSeriesProposal: Record "No. Series Proposal"; var ResponseText: Text; var NoSeriesGenerated: Record "No. Series Proposal Line"; InputText: Text)
```
GeneratedNoSeriesProposalLine? maybe with some abbr. ?

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+        end;
+    end;
+
+    procedure ApplyProposedNoSeries(var NoSeriesGenerated: Record "No. Series Proposal Line")
```
same, i'd like the variable name to somewhat match the object name. Variables can add descriptors to make it more precise but the more they differ the more I think we simply need better names...

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+    var
+        NoSeries: Record "No. Series";
+    begin
+        NoSeries.Init();
```
NoSeries is a local record variable and not in a loop so there is nothing to Init ...

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+    var
+        NoSeriesLine: Record "No. Series Line";
+    begin
+        NoSeriesLine.Init();
```
same, don't init when not needed.

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+        NoSeries.Description := NoSeriesGenerated.Description;
+        NoSeries."Manual Nos." := true;
+        NoSeries."Default Nos." := true;
+        //TODO: Check if we need to add more fields here, like "Mask", "No. Series Type", "Reverse Sales VAT No. Series" etc.
```
having todo's in code worries me... is this just a note? should it be fixed before we ship? What if we don't fix it? 

I dislike comments in code because they get outdated very quickly but I can live with them (docs are better :P) but todo's like this should not be in code.

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+        NoSeriesLine."Ending No." := NoSeriesGenerated."Ending No.";
+        NoSeriesLine."Warning No." := NoSeriesGenerated."Warning No.";
+        NoSeriesLine."Increment-by No." := NoSeriesGenerated."Increment-by No.";
+        //TODO: Check if we need to add more fields here, like "Allow Gaps in Nos.", "Sequence Name" etc.
```
I think it's important to set the "Allow Gaps" / Implementation field which depends on the purpose of the no. series. 
I think we default to Normal / no gaps because it's the safe choice but we prefer Sequence / allow gaps because it has better perf.

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+        NoSeries: Codeunit "No. Series";
+    begin
+        if not NoSeries.GetNoSeriesLine(NoSeriesLine, SeriesCode, 0D, true) then
+            exit(1000);
```
```suggestion
            exit(10000);
```

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+        FieldRef: FieldRef;
+    begin
+        RecRef.Open(NoSeriesGenerated."Setup Table No.");
+        if not RecRef.FindFirst() then
```
Since the assumption here seems to be that there is only 1 setup record and that record would have a blank primary key,  should we not assert that more aggressively? Instead of finding the first record maybe we should 'get' the record with the blank prim. key. ... or check that the record we find has a blank value.

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Setup [DRAFT]/NoSeriesCopilotSetup.Table.al:
```diff
@@ -0,0 +1,679 @@
+        }
+    }
+
+    procedure GetEndpoint() Endpoint: Text[250]
```
All these methods should be add to a codeunit somewhere. Logic should reside in codeunits not tables, pages etc...

I get the appeal of these getters but they are not used consistently. Sometimes they are used, sometimes not. They are also not using the Rec. as I'd expect but instead do their own Get... yes, these will likely hit the cache but still, we have Setup.Get all over the place now instead of getting the record once only..

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+            AOAIChatMessages.DeleteMessage(AOAIChatMessages.GetHistory().Count); // remove the last message, as it is not needed anymore
+            AOAIChatMessages.DeleteMessage(AOAIChatMessages.GetHistory().Count); // remove the tools message, as it is not needed anymore
+
+            Sleep(1000); // sleep for 1000ms, as the model can be called only limited number of times per second
```
is 1000ms really the right number? do we have any plan to finetune this a little more?

darjoo replied:
Removed

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,518 @@
+
+    local procedure MaxModelTokens(): Integer
+    begin
+        exit(16385); //gpt-3.5-turbo-1106
```
should this be updated? I thought I saw some call to gpt 4 higher up...

grobyns commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesProposal.Page.al:
```diff
@@ -0,0 +1,197 @@
+            ApplyProposedNoSeries();
+    end;
+
+    local procedure GenerateNoSeries()
```
these methods should be moved to a codeunit... and we should have a oneliner here at most..

MPGoldenEDI commented:
Proposing numerical sequences for use in production settings diverges from those in demonstrations. It's important to evaluate the long-term viability of each series and distinguish between various types of documents.
My experience is that the starting point never takes into concideration that the series must last 20 year maybe more. 
My suggestion is to include the year and then have a length that covers a growth in the company. A prefix to the year can include alphanumeric characters, also inposted documents, if thats acceptabel by the company. Else the suggestion is to hav a number as prefix on documents that is used for external communication.
The advantage is that each document number clearly indicates the year of creation to the user, and the sorting sequence will invariably follow a logical order.
Example: 
Sales Order:
a) SO24000001 when alphanumeric is allowed
b) 124000001 when alphanumeric is not allowed
Posted Sales Invoice:
a) SI24000001 when alphanumeric is allowed
b) 224000001 when alphanumeric is not allowed

I have seen so many systems where the partner has started with Contoso/Cronus number series and belive that this is the way to go. 
I have helped customers reconstruct their number series, yet they still struggle with poor sorting.

DmitryKatson commented:
> Here are my notes/questions I have from our sync :)
> 

@darjoo 
> * Action visibility
>   
>   * Addition to SaaS check, we need to limit CA and potentially any other languages. This is not necessary to add now but 
I'll add the code snippets below. CA is due to language requirement that both EN & FR are supported. Languages wise, we've typically stuck with EN first as Production Ready Previews and slowly roll out other languages as the language assessment team gives us the green light
>

Added `codeunit 324 "No. Series Copilot Impl.".IsCopilotVisible()` similar to other copilots you have in the system. It checks SaaS environment (deactivated for now, as I work in Docker) and supported languages (similar to other copilots, only -pl is excluded).

> * Use prompt guides
>   
>   * For the prompt dialog, using prompt guides can give the users an easy way to select premade prompts. The ones that you think are most likely to be used by them. I'm not sure if the docs are ready yet about them, but they will be mentioned in the Launch Event for What's New: Extending Copilot with AL
>

It has prompt guides and a placeholder now

> * There's a check for whether input against the max tokens, and it just exits if it's greater than the max tokens allowed. We should at least give an actionable error or see if it can be modified to always allow even if it's multiple requests
>

I added the `codeunit 348 "No. Ser. Cop. Notific. Manager"` and used `NotificationManager.SendNotification(ChatCompletionResponseErr);` in case of out of tokens scenario and also inside of the functions implementations codeunits - similar as you do in the `Sales Lines Suggestions Impl.`

> * When the generation comes back with a response, would it be beneficial for `AOAIChatMessages` to give you say, `GetLastTool` or `GetLastFunction` when it's a tool/function and provide you the `Name:Text; Arguments: JsonArray` (for example) instead of having to parse `GetLastMessage`?
> * I'll add a method in `AOAIChatMessages` to clear the existing tools. So you don't have to loop through and delete all tools.
>

Since I changed the approach from "Get Intent > Run intent function > Add result to chat history > get final answer" to "Get Intent > Run intent function > Get Result > Make standalone AI call without chat history" this is not relevant anymore. But thanks for reimplemening the function calling. Much easier to work now. 

> * In your `BuildGenerateNewNumberSeries` function, it runs tokencount for each line. I'll be improving that, so you can put in `AOAIChatMessages` directly and it'll count everything that's being sent.
>

I've removed token counting inside of the functions calling. Instead, it's chunking requests based on number series to generate. As now i use standalone calls, instead of sending chat history from the beginning and I build the final prompt, I'm pretty in control of used tokens. According to the testing each call consumes average 15% of prompt tokens and 50% of completion tokens.

> * Filtering on name with `* Setup` for tables is risky as we can't guarantee they always end with `Setup`

There was a discussion on this, but seems we don't have other ways, currently how to find setup tables. Some additional checks could be that
- There is only 1 record
- There is a Code field
- Code field is Blank
Probably this will be a documented mandatory requirement to make this copilot work.  But ideally, we should have TableType::Setup.

> * GPT3.5 has a context window of 16k while GPT4 has 32k and turbo 128k, would it be better to switch?
I tested with GPT4 and GPT4-o. But currently GPT3.5 gives fastest and reliable responses with a very hand-crafted prompts that I use, at least on the test questions. 
>

>   
>   * This you had answered already, hallucination can be pretty bad with generating more tables. So there isn't a point in using GPT4
> * Does generation of ~10 tables always work?

I followed your advice, and instead of "ask in the next Json format" approach I use now 3rd function and get Json array as input. This was a really good advice, which allowed to increase correctness from 10 tables to 50. However, in final version of the copilot I left 40 tables, as 50 works in 9 cases out of 10, according to my tests. And 40 seems to work always and usually cover most of the scenarios. 

>   
>   * This you had also answered, it doesn't as such it does retries and completely fails if it doesn't generate all.
>   * We should allow partial generation and then allow the the user to request another generation on the ungenerated ones, or a full regenerate.
>

This is a good idea, but not in this version. However, after moving to function call, instead of json output, and additional prompt finetuning i see much reduce in wrong generations. We also discussed this, outside of this thread, and decided to add `"No. Series Copilot Impl.".MinimumAccuracy()` which is `0.9` for now. It means that if user requested to generate number series for 40 tables and only 36 will be generated, then we keep this result and don't regenerate. 

Also, i implemented a logic, that if number series is already set up in the "[area] Setup" table, then we don't need to generate it once again. Meaning, that if user asks "Generate me number series for purchase module", accepts the first result and then asks once again, there will be no generations as it's already configured. This will reduce the number of "duplications" errors and also quality will increase.

> * How long does it take to run? As running 10+ requests will take time, including retries that makes it worse.
>   
>   * Takes about a minute. This might be fine as this is an admin scenario where they're setting the system up. And this saves them a lot more time in the long run.
>

My tests say that approx. 38 seconds to generate 40 number series.

> * Retry logic, what if 1 out of 10 fails?
>   
>   * Similar qs to the generation of ~10 tables always work. You mentioned that all of them fail. We should allow partial generation if it makes sense.
> 

Answered above

> Thank you so much for going through the code and all! Looking forward to when this sees the light of day in the product!

Thank you for support in this journey ;)

DmitryKatson commented:
😅 "I worked really hard (c)" and now we have improved version of No Series Copilot. 

1) It's more stable in terms of generated results and require maximum 3 calls to AOAI instead of original 12 to generate no series for blank company. The generation time is still same, We need to generate same number of number series, so tokens/sec speed is a cumber stone here. 
2) The total number of tokens require to generate no series for blank company, reduced apprx. 60%. 

This was achieved, due to new way of generating the final json array of number series. Thanks @darjoo for advice.
Instead of just asking LLM to "Please generate me a number series in a Json Array of next format..." we add a function with parameter of JsonArray type. We ask LLM to generate for us function parameter. Appeared it works much better, with slight prompting adjustments, at least X5 increase in result quality.

JesperSchulz reviewed (changes requested):
We should look into the remaining TODOs and open code review comments. It's time to get this code production ready. When we think we're there, we can get one final round of review from the product group.

JesperSchulz commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,540 @@
+        GeneratingNoSeriesForLbl: Label 'Generating number series %1', Comment = '%1 = No. Series';
+        FeatureNameLbl: Label 'Number Series with AI', Locked = true;
+
+    internal procedure GetNoSeriesSuggestions()
```
Given the entire codeunit is internal, the procedures don't have to be.

JesperSchulz commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesProposal.Table.al:
```diff
@@ -0,0 +1,50 @@
+
+namespace Microsoft.Foundation.NoSeries;
+
+table 391 "No. Series Proposal"
```
I'm not a fan of this name. Let's find something better.

DmitryKatson commented:
🧨 New Skill is added: Prepare for next year.
This skill (or tool) allows to create new number series lines for the next year, if number series are yearly based (have year prefix in the number).

See in the video example. 

https://github.com/microsoft/BCApps/assets/29101017/543978d9-3294-44da-8e60-8510420b8515

darjoo commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,565 @@
+    var
+    // EnvironmentInformation: Codeunit "Environment Information";
+    begin
+        // if not EnvironmentInformation.IsSaaSInfrastructure() then //TODO: Check how to keep IsSaaSInfrastructure but be able to test in Docker Environment
```
Todo, put these back

darjoo commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al:
```diff
@@ -0,0 +1,565 @@
+        LanguageSelection.SetLoadFields("Language Tag");
+        LanguageSelection.SetRange("Language ID", UserSessionSettings.LanguageId());
+        if LanguageSelection.FindFirst() then
+            if LanguageSelection."Language Tag".StartsWith('pt-') then
```
@DenLilleMand We can remove this right? We no longer have the portugese limitation for any feature.

DenLilleMand replied:
@darjoo yes we should get rid of everything that filters Copilot features based on language or regions(e.g. Canada was denied access earlier)

darjoo commented on src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesProposal.Page.al:
```diff
@@ -0,0 +1,231 @@
+                action(SetupModuleForNextYear)
+                {
+                    Caption = 'Set up number series for the [sales] module for the next year';
+                    ToolTip = 'Sample prompt for setting up number series for a specific module for the next year. Replace [sales] with the module you want to set up number series for.';
```
Todo on Darrick. To get design folks to review tooltips & captions.

darjoo replied:
To do after merge

darjoo commented on src/Business Foundation/App/NoSeriesCopilot/src/Register/NoSeriesCopilotRegister.Codeunit.al:
```diff
@@ -0,0 +1,46 @@
+    end;
+
+    var
+        LearnMoreUrlTxt: Label 'https://go.microsoft.com/fwlink/?linkid=[id]', Locked = true; //TODO: replace 'id' when docs page is ready
```
Todo for Darrick. Work with Brent to get a docs page id for this.

darjoo replied:
To do after merge

darjoo commented on src/Business Foundation/App/NoSeriesCopilot/src/Setup [DRAFT]/NoSeriesCopilotSetup.Table.al:
```diff
@@ -0,0 +1,287 @@
+
+namespace Microsoft.Foundation.NoSeries;
+
+table 9200 "No. Series Copilot Setup"
```
This will be removed.

darjoo commented on src/Business Foundation/App/NoSeriesCopilot/src/Setup [DRAFT]/NoSeriesCopilotSetup.Page.al:
```diff
@@ -0,0 +1,200 @@
+/// <summary>
```
This will be removed

darjoo commented on src/Business Foundation/App/NoSeriesCopilot/src/Utilities/NoSerCopNotificManager.Codeunit.al:
```diff
@@ -0,0 +1,33 @@
+// ------------------------------------------------------------------------------------------------
+namespace Microsoft.Foundation.NoSeries;
+
+codeunit 348 "No. Ser. Cop. Notific. Manager"
```
Move these into the `No Series Copilot Impl`. Doesn't make sense for a separate codeunit for these functions.
